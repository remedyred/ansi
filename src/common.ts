export type CodePair = [number, number]

export type Modifier = 'bold' | 'dim' | 'hidden' | 'inverse' | 'italic' | 'overline' | 'reset' | 'strikethrough' | 'underline'

export type Color = 'black' | 'blackBright' | 'blue' | 'blueBright' | 'cyan' | 'cyanBright' | 'gray' | 'green' | 'greenBright' | 'grey' | 'magenta' | 'magentaBright' | 'red' | 'redBright' | 'white' | 'whiteBright' | 'yellow' | 'yellowBright'

export type BackgroundColor = 'bgBlack' | 'bgBlackBright' | 'bgBlue' | 'bgBlueBright' | 'bgCyan' | 'bgCyanBright' | 'bgGray' | 'bgGreen' | 'bgGreenBright' | 'bgGrey' | 'bgMagenta' | 'bgMagentaBright' | 'bgRed' | 'bgRedBright' | 'bgWhite' | 'bgWhiteBright' | 'bgYellow' | 'bgYellowBright'

export type StyleCategory<Allowed extends string> = Partial<AnsiControls> & Partial<AnsiMethods> & Record<Allowed, CodePair>

export type AnsiRenderFn = (...args) => string

export type AnsiMethods = {
	ansi: AnsiRenderFn
	ansi256: AnsiRenderFn
	ansi16m: AnsiRenderFn
}

export interface AnsiControls {
	close: string
	open: string
}

export interface AnsiCodes {
	modifier: Record<Modifier, CodePair>
	color: Record<Color, CodePair>
	bgColor: Record<BackgroundColor, CodePair>
}

export interface StyleCodes {
	modifier: StyleCategory<Modifier>
	color: StyleCategory<Color>
	bgColor: StyleCategory<BackgroundColor>
}

export interface Styles extends StyleCodes {
	rgbToAnsi256(r: number, g: number, b: number): number
	hexToRgb(hex: string): [number, number, number]
	ansi256ToAnsi(color: number): number
	hexToAnsi256(hex: string): number
	hexToAnsi(hex: string): number
	rgbToAnsi(red, green, blue): number
}

export const ANSI_BACKGROUND_OFFSET = 10

// Setup gray and bgGray since they are used multiple times
const gray: CodePair = [90, 39]
const bgGray: CodePair = [100, 49]

export const ansiCodes: AnsiCodes = {
	modifier: {
		reset: [0, 0],
		// 21 isn't widely supported and 22 does the same thing
		bold: [1, 22],
		dim: [2, 22],
		italic: [3, 23],
		underline: [4, 24],
		overline: [53, 55],
		inverse: [7, 27],
		hidden: [8, 28],
		strikethrough: [9, 29]
	},
	color: {
		black: [30, 39],
		red: [31, 39],
		green: [32, 39],
		yellow: [33, 39],
		blue: [34, 39],
		magenta: [35, 39],
		cyan: [36, 39],
		white: [37, 39],
		gray,
		grey: gray,

		// Bright color
		blackBright: gray,
		redBright: [91, 39],
		greenBright: [92, 39],
		yellowBright: [93, 39],
		blueBright: [94, 39],
		magentaBright: [95, 39],
		cyanBright: [96, 39],
		whiteBright: [97, 39]
	},
	bgColor: {
		bgBlack: [40, 49],
		bgRed: [41, 49],
		bgGreen: [42, 49],
		bgYellow: [43, 49],
		bgBlue: [44, 49],
		bgMagenta: [45, 49],
		bgCyan: [46, 49],
		bgWhite: [47, 49],
		bgGray,
		bgGrey: bgGray,

		// Bright color
		bgBlackBright: bgGray,
		bgRedBright: [101, 49],
		bgGreenBright: [102, 49],
		bgYellowBright: [103, 49],
		bgBlueBright: [104, 49],
		bgMagentaBright: [105, 49],
		bgCyanBright: [106, 49],
		bgWhiteBright: [107, 49]
	}
}
