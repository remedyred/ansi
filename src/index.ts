export * from './ansi-regex'
export * from './ansi-styles'
export * from './strip-ansi'
