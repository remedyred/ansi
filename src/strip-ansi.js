import {ansiRegex} from './ansi-regex'

export function stripAnsi(string) {
	if (typeof string !== 'string') {
		throw new TypeError(`Expected a \`string\`, got \`${typeof string}\``)
	}

	return string.replace(ansiRegex(), '')
}
