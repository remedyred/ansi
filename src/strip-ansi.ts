import {ansiRegex} from './ansi-regex'

/**
 * Remove ansi escape codes from a string
 * @param text
 */
export function stripAnsi(text: string): string {
	if (typeof text !== 'string') {
		throw new TypeError(`Expected a \`string\`, got \`${typeof text}\``)
	}

	return text.replace(ansiRegex(), '')
}
