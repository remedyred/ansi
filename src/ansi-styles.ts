import {ANSI_BACKGROUND_OFFSET, ansiCodes, AnsiCodes, AnsiRenderFn, StyleCodes, Styles} from './common'

function wrapAnsi16(offset = 0): AnsiRenderFn {
	return (code: number): string => {
		return `\u001B[${code + offset}m`
	}
}

function wrapAnsi256(offset = 0): AnsiRenderFn {
	return (code: number): string => {
		return `\u001B[${38 + offset};5;${code}m`
	}
}

function wrapAnsi16m(offset = 0): AnsiRenderFn {
	return (red: number, green: number, blue: number): string => {
		return `\u001B[${38 + offset};2;${red};${green};${blue}m`
	}
}

function generateStyleCodes() {
	const codes = new Map()

	const styleCodes: StyleCodes = {...ansiCodes}

	for (const [groupName, group] of Object.entries(styleCodes)) {
		for (const [styleName, codePair] of Object.entries(group)) {
			styleCodes[styleName] = {
				open: `\u001B[${codePair[0]}m`,
				close: `\u001B[${codePair[1]}m`
			}

			group[styleName] = styleCodes[styleName]

			codes.set(codePair[0], codePair[1])
		}

		Object.defineProperty(styleCodes, groupName, {
			value: group,
			enumerable: false
		})
	}

	Object.defineProperty(styleCodes, 'codes', {
		value: codes,
		enumerable: false
	})

	styleCodes.color.close = '\u001B[39m'
	styleCodes.bgColor.close = '\u001B[49m'

	styleCodes.color.ansi = wrapAnsi16()
	styleCodes.color.ansi256 = wrapAnsi256()
	styleCodes.color.ansi16m = wrapAnsi16m()
	styleCodes.bgColor.ansi = wrapAnsi16(ANSI_BACKGROUND_OFFSET)
	styleCodes.bgColor.ansi256 = wrapAnsi256(ANSI_BACKGROUND_OFFSET)
	styleCodes.bgColor.ansi16m = wrapAnsi16m(ANSI_BACKGROUND_OFFSET)

	return styleCodes
}

function styleFactory(): Styles {
	const styles = generateStyleCodes() as Styles

	// From https://github.com/Qix-/color-convert/blob/3f0e0d4e92e235796ccb17f6e85c72094a651f49/conversions.js
	Object.defineProperties(styles, {
		rgbToAnsi256: {
			value(red, green, blue) {
				// We use the extended greyscale palette here, with the exception of
				// black and white. normal palette only has 4 greyscale shades.
				if (red === green && green === blue) {
					if (red < 8) {
						return 16
					}

					if (red > 248) {
						return 231
					}

					return Math.round((red - 8) / 247 * 24) + 232
				}

				return 16 +
					36 * Math.round(red / 255 * 5) +
					6 * Math.round(green / 255 * 5) +
					Math.round(blue / 255 * 5)
			},
			enumerable: false
		},
		hexToRgb: {
			value(hex) {
				const matches = /(?<colorString>[\da-f]{6}|[\da-f]{3})/i.exec(hex.toString(16))
				if (!matches) {
					return [0, 0, 0]
				}

				let {colorString} = matches.groups

				if (colorString.length === 3) {
					colorString = colorString.split('').map(character => character + character)
						.join('')
				}

				const integer = Number.parseInt(colorString, 16)

				return [integer >> 16 & 0xFF, integer >> 8 & 0xFF, integer & 0xFF]
			},
			enumerable: false
		},
		hexToAnsi256: {
			value: hex => styles.rgbToAnsi256(...styles.hexToRgb(hex)),
			enumerable: false
		},
		ansi256ToAnsi: {
			value(code) {
				if (code < 8) {
					return 30 + code
				}

				if (code < 16) {
					return 90 + (code - 8)
				}

				let red
				let green
				let blue

				if (code >= 232) {
					red = ((code - 232) * 10 + 8) / 255
					green = red
					blue = red
				} else {
					code -= 16

					const remainder = code % 36

					red = Math.floor(code / 36) / 5
					green = Math.floor(remainder / 6) / 5
					blue = remainder % 6 / 5
				}

				const value = Math.max(red, green, blue) * 2

				if (value === 0) {
					return 30
				}

				let result = 30 + (Math.round(blue) << 2 | Math.round(green) << 1 | Math.round(red))

				if (value === 2) {
					result += 60
				}

				return result
			},
			enumerable: false
		},
		rgbToAnsi: {
			value: (red, green, blue) => styles.ansi256ToAnsi(styles.rgbToAnsi256(red, green, blue)),
			enumerable: false
		},
		hexToAnsi: {
			value: hex => styles.ansi256ToAnsi(styles.hexToAnsi256(hex)),
			enumerable: false
		}
	})

	return styles
}

export {Styles, StyleCategory, BackgroundColor, Color, Modifier, AnsiControls, AnsiMethods, CodePair, AnsiRenderFn} from './common'
export const ansiStyles = styleFactory()
