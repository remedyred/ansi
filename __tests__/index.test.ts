import {ansiRegex, ansiStyles, generateStyleCodes, stripAnsi} from '../src'

const BLACK = '\u001b[30m'
const RED = '\u001b[31m'
const GREEN = '\u001b[32m'
const ANSI_STRING = `${RED} ${GREEN} ${BLACK}`


describe("ansiRegex", () => {
	it('should return an instance of RegExp', () => expect(ansiRegex()).toBeInstanceOf(RegExp))

	it('should have the g flag', () => expect(ansiRegex().flags).toBe('g'))

	it('(true) should NOT have the g flag', () => expect(ansiRegex(true).flags).toBe(''))

	it('should return the ansi string', () => {
		expect(ANSI_STRING.match(ansiRegex())).toStrictEqual([RED, GREEN, BLACK])
	})

	it('(true) should return the first ansi string', () => {
		expect(ansiRegex(true).exec(ANSI_STRING)[0]).toBe(RED)
	})
})

describe("stripAnsi", () => {
	it('should strip ansi from a string', () => {
		expect(stripAnsi(ANSI_STRING)).toBe('  ')
	})
})

describe('ansiStyles', () => {
	it('should have a color property', () => expect(ansiStyles.color).toBeDefined())
	it('should have a bgColor property', () => expect(ansiStyles.bgColor).toBeDefined())
	it('should have a modifier property', () => expect(ansiStyles.modifier).toBeDefined())
	it('should have a rgbToAnsi256 method', () => expect(ansiStyles.rgbToAnsi256).toBeInstanceOf(Function))
	it('should have a hexToRgb method', () => expect(ansiStyles.hexToRgb).toBeInstanceOf(Function))
	it('should have a hexToAnsi256 method', () => expect(ansiStyles.hexToAnsi256).toBeInstanceOf(Function))
	it('should have a ansi256ToAnsi method', () => expect(ansiStyles.ansi256ToAnsi).toBeInstanceOf(Function))
	it('should have a rgbToAnsi method', () => expect(ansiStyles.rgbToAnsi).toBeInstanceOf(Function))
	it('should have a hexToAnsi method', () => expect(ansiStyles.hexToAnsi).toBeInstanceOf(Function))
})

describe("generateStyleCodes", () => {
	it("generateStyleCodes should match snapshot", () => {
		expect(generateStyleCodes()).toMatchSnapshot()
	})
})
