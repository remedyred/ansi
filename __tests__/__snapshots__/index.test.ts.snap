// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateStyleCodes generateStyleCodes should match snapshot 1`] = `
Object {
  "ansi": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "ansi16m": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "ansi256": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "bgBlack": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "bgBlackBright": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "bgBlue": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "bgBlueBright": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "bgCyan": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "bgCyanBright": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "bgGray": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "bgGreen": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "bgGreenBright": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "bgGrey": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "bgMagenta": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "bgMagentaBright": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "bgRed": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "bgRedBright": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "bgWhite": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "bgWhiteBright": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "bgYellow": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "bgYellowBright": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "black": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "blackBright": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "blue": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "blueBright": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "bold": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "close": Object {
    "close": "[[m",
    "open": "[m",
  },
  "cyan": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "cyanBright": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "dim": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "gray": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "green": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "greenBright": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "grey": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "hidden": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "inverse": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "italic": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "magenta": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "magentaBright": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "overline": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "red": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "redBright": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "reset": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "strikethrough": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "underline": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "white": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "whiteBright": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "yellow": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
  "yellowBright": Object {
    "close": "[undefinedm",
    "open": "[undefinedm",
  },
}
`;
